src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs

```diff
@@ -197,6 +203,44 @@ public override string ToString()
             return _uri.ToString();
         }
 
+        private void EncodeValue(string value)
+        {
+            EncodeValue(value, encodeSlashes: true);
+        }
+
+        private void EncodeValue(string value, bool encodeSlashes)
+        {
+            EncodeValue(value, startIndex: 0, characterCount: value.Length, encodeSlashes);
+        }
+
+        // For testing
+        internal void EncodeValue(string value, int startIndex, int characterCount, bool encodeSlashes)
+        {
+            // Just encode everything if its ok to encode slashes
+            if (encodeSlashes)
+            {
+                _urlEncoder.Encode(Writer, value, startIndex, characterCount);
+            }
+            else
+            {
+                var length = startIndex + characterCount;
+                var lastStartIndex = startIndex;
+                for (var i = startIndex; i < length; i++)
```

Comment:
nit: I would expect this loop to look more like: https://github.com/aspnet/Routing/blob/release/2.2/src/Microsoft.AspNetCore.Routing/Matching/FastPathTokenizer.cs#L29

# CHANNEL YOUR INNER C PROGRAMMER